pkg/kubefed2/disable.go

```diff
@@ -139,75 +167,184 @@ func (j *disableType) Run(cmdOut io.Writer, config util.FedConfig) error {
 		Namespace: j.FederationNamespace,
 		Name:      name,
 	}
-
-	return DisableFederation(cmdOut, hostConfig, typeConfigName, j.delete, j.DryRun)
+	j.enableTypeDirective.Name = typeConfigName.Name
+	return DisableFederation(cmdOut, hostConfig, j.enableTypeDirective, typeConfigName, j.deleteCRD, j.DryRun)
 }
 
-func DisableFederation(cmdOut io.Writer, config *rest.Config, typeConfigName ctlutil.QualifiedName, delete, dryRun bool) error {
+func DisableFederation(cmdOut io.Writer, config *rest.Config, enableTypeDirective *enable.EnableTypeDirective,
+	typeConfigName ctlutil.QualifiedName, deleteCRD, dryRun bool) error {
 	client, err := genericclient.New(config)
 	if err != nil {
 		return errors.Wrap(err, "Failed to get federation clientset")
 	}
+
+	write := func(data string) {
+		if cmdOut == nil {
+			return
+		}
+
+		if _, err := cmdOut.Write([]byte(data)); err != nil {
+			glog.Fatalf("Unexpected err: %v\n", err)
+		}
+	}
+
 	typeConfig := &fedv1a1.FederatedTypeConfig{}
-	err = client.Get(context.TODO(), typeConfig, typeConfigName.Namespace, typeConfigName.Name)
+	ftcExists, err := checkFederatedTypeConfigExists(client, typeConfig, typeConfigName, write)
 	if err != nil {
-		return errors.Wrapf(err, "Error retrieving FederatedTypeConfig %q", typeConfigName)
+		return err
 	}
 
 	if dryRun {
 		return nil
 	}
 
-	write := func(data string) {
-		if cmdOut == nil {
-			return
+	if deleteCRD {
+		if !ftcExists {
+			typeConfig, err = generatedFederatedTypeConfig(config, enableTypeDirective)
+			if err != nil {
+				return err
+			}
 		}
+		err = deleteFederatedType(config, typeConfig, write)
```

Comment:
I'm recalling the original reason why the propagation was being disabled.  Deleting the crd while propagation is still being performed is likely to generate a bunch of spurious error messages in the controller manager log as the watches are terminated and can't be reestablished.  Deleting the FTC before the CRD risks breaking idempotent behavior if deletion of the CRD is subsequently unsuccessful, since a subsequent disable attempt may not be able to find the CRD again (if it has non-standard group/version/name/etc).  Disabling, deleting CRD, then deleting FTC would seem the cleanest path.

Reply:
I see, makes sense. Updated.